src/main/frontend/fs/sync.cljs

```diff
@@ -0,0 +1,1223 @@
+              (if (instance? ExceptionInfo diff-r)
+                diff-r
+                (let [[diff-txns latest-txid min-txid] diff-r]
+                  (if (> (dec min-txid) @*txid) ;; if min-txid-1 > @*txid, need to remote->local-full-sync
+                    (do (println "min-txid" min-txid "request-txid" @*txid)
+                        {:need-remote->local-full-sync true})
+
+                    (when (pos-int? latest-txid)
+                      (let [partitioned-filetxns (transduce (diffs->partitioned-filetxns 10)
+                                                            (completing (fn [r i] (conj r (reverse i)))) ;reverse
+                                                            '()
+                                                            (reverse diff-txns))]
+                        ;; (prn "partition-filetxns" partitioned-filetxns)
+
+                        ;; TODO: precheck etag
+                        (if (empty? (flatten partitioned-filetxns))
+                          (do (update-graphs-txid! latest-txid graph-uuid repo)
+                              (reset! *txid latest-txid)
+                              {:succ true})
+                          (<! (apply-filetxns-partitions
+                               *sync-state graph-uuid base-path partitioned-filetxns repo *txid *stopped)))))))))]
+        (cond
+          (instance? ExceptionInfo r)
+          {:unknown r}
+
+          @*stopped
+          {:stop true}
+
+          (:need-remote->local-full-sync r)
+          r
+
+          :else
+          {:succ true}))))
+
+  (sync-remote->local-all-files! [this]
+    (go
+      (let [remote-all-files-meta-c (get-remote-all-files-meta remoteapi graph-uuid)
+            local-all-files-meta-c (get-local-all-files-meta rsapi graph-uuid base-path)
+            remote-all-files-meta (<! remote-all-files-meta-c)
+            local-all-files-meta (<! local-all-files-meta-c)
+            diff-remote-files (set/difference remote-all-files-meta local-all-files-meta)
+            latest-txid (:TXId
+                         (<! (get-remote-graph remoteapi nil graph-uuid)))]
+        (println "[full-sync(remote->local)]"
+                 (count diff-remote-files) "files need to sync")
+        (<! (.sync-files-remote->local!
+             this (map -relative-path diff-remote-files)
+             latest-txid))))))
+
+(defn- file-changed?
+  "return true when file changed compared with remote"
+  [graph-uuid file-path-without-base-path base-path]
+  (go
+    (let [remote-meta (first (<! (get-remote-files-meta remoteapi graph-uuid [file-path-without-base-path])))
+          local-meta (first (<! (get-local-files-meta rsapi graph-uuid base-path [file-path-without-base-path])))]
+      (not= remote-meta local-meta))))
+
+(defn- contains-path? [regexps path]
+  (reduce #(when (re-find %2 path) (reduced true)) false regexps))
+
+
+(deftype ^:large-vars/cleanup-todo
```

Comment:
Moving out a protocol fn or two could be helpful. Having longer fns makes it likely you'll be the only one on the team who can make changes here quickly. One example is [query-dsl/build](https://github.com/logseq/logseq/blob/a175b5168409b7f21799214aedbc20c912776cdc/src/main/frontend/db/query_dsl.cljs#L145-L388) which only [tienson made changes on for over a year](https://github.com/logseq/logseq/commits/master?before=d74f75d1a16cc14224bcd022d24d9c29d60f7e1c+35&branch=master&path%5B%5D=src&path%5B%5D=main&path%5B%5D=frontend&path%5B%5D=db&path%5B%5D=query_dsl.cljs) despite it having bugs